src/state/block_tree.rs

```diff
@@ -0,0 +1,484 @@
+            Ok(self.block_height(&block)?)
+        } else {
+            Ok(None)
+        }
+    }
+
+    /* ↓↓↓ WriteBatch commit ↓↓↓ */
+
+    pub fn write(&mut self, write_batch: BlockTreeWriteBatch<K::WriteBatch>) {
+        self.0.write(write_batch.0)
+    }
+
+    /* ↓↓↓ Snapshot ↓↓↓ */
+
+    pub fn snapshot(&self) -> BlockTreeSnapshot<K::Snapshot<'_>> {
+        BlockTreeSnapshot::new(self.0.snapshot())
+    }
+
+    /* ↓↓↓ Get AppBlockTreeView for ProposeBlockRequest and ValidateBlockRequest */
+
+    pub fn app_view<'a>(
+        &'a self,
+        parent: Option<&CryptoHash>,
+    ) -> Result<AppBlockTreeView<'a, K>, BlockTreeError> {
+        let highest_committed_block_height = self.highest_committed_block_height()?;
+        let parent = match parent {
+            None => None,
+            Some(&b) => Some(b),
+        };
+
+        // Obtain an iterator over the ancestors starting from the parent, all the way until genesis,
+        // from newest (parent) to oldest.
+        let ancestors_iter = successors(parent, |b| {
+            self.block_justify(b)
+                .ok()
+                .map(|qc| {
+                    if !qc.is_genesis_qc() {
+                        Some(qc.block)
+                    } else {
+                        None
+                    }
+                })
+                .flatten()
+        });
+
+        let ancestors_heights_iter = ancestors_iter
+            .clone()
+            .map(|block| {
+                self.block_height(&block).map(|res| {
+                    if res.is_none() {
+                        Err(BlockTreeError::BlockExpectedButNotFound {
+                            block: block.clone(),
+                        })
+                    } else {
+                        Ok(res.unwrap())
+                    }
+                })
+            })
+            .flatten()
+            .flatten();
+
+        // Obtain an iterator over the uncomitted ancestors starting from the parent,
+        // ending at the lowest uncommitted ancestor.
+        let uncommitted_ancestors_iter = ancestors_iter
+            .zip(ancestors_heights_iter)
+            .take_while(|(_, height)| {
+                highest_committed_block_height.is_none()
+                    || highest_committed_block_height.is_some_and(|h| height > &h)
+            })
+            .map(|(b, _)| b);
+
+        // Obtain a vector of optional app state updates associated with ancestors
+        // starting from the parent, ending at the oldest uncommitted ancestor.
+        let pending_ancestors_app_state_updates: Vec<Option<AppStateUpdates>> =
+            uncommitted_ancestors_iter
+                .map(|block| self.pending_app_state_updates(&block))
+                .flatten()
+                .collect();
+
+        Ok(AppBlockTreeView {
+            block_tree: self,
+            pending_ancestors_app_state_updates,
+        })
+    }
+
+    /* ↓↓↓ Basic state getters ↓↓↓ */
```

Comment:
The return values of these "basic state getters" are inconsistent. Some of them return a `Result`, some of them don't. Some of them return an `Option`, and some of them don't.

Reply:
Will fix with https://github.com/parallelchain-io/hotstuff_rs/issues/44